### ISSUE TITLE
Clang 15: -Wunused-but-set-variable doesn't recognize pre- or post-increment as "use"

### ISSUE THREAD

**Issue by ednolan** [llvm/llvm-project#58426]
This seems to be a Clang 15 regression; the provided example works in Clang 13 and 14.

This code snippet:
```
void foo() {
    int bar = 0;
    ++bar;
}
```

With these command line parameters:
`-Wunused-but-set-variable -Werror -c`

Results in this compilation error:
```
<source>:2:9: error: variable 'bar' set but not used [-Werror,-Wunused-but-set-variable]
    int bar = 0;
        ^
```

Compiler explorer link:
https://godbolt.org/z/dKW5qhroz

The same issue occurs if you write `bar++` instead of `++bar`. It goes away if you write `bar = bar + 1`.

**Comment by LebedevRI** [comment]
The diagnostic is correct, pre- or post-increment is not an use.
In fact, not diagnosing `bar = bar + 1` is a bug.

**Comment by dwblaikie** [comment]
Not sure I'd say that not diagnosing `bar = bar + 1` is a bug as such, but a false negative/missing case - it'd require more complex expression analysis to diagnose, which we don't do yet - not that we couldn't do, but might not be worth the tradeoff, etc.

But yeah, for expressions where it's trivially obvious "this is using the variable, but in no way that's observable" (like pre/post increment) we don't warn. Differentiating `bar = bar + 1` from `baz = bar + 1` or `bar = func_with_side_effects(bar)` or `bar = (++baz) + bar`, etc... would be a bit of work/hasn't been done yet.
